### PR TITLE
Briefly document dominator analysis

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a5238b60f263dae202a74d5bc394f66a22daf27f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91253997

File: src/analyses/README.md [hunk 1]
```diff
@@ -42,7 +42,26 @@ that work with the \ref grapht representation.
...
+This is useful for e.g. checking against introducing a redundant check: if we
+have a `const int *x` parameter and wish to introduce a null check at node Y,
+then we can also mark it checked at all the nodes that Y dominates, as at those
+program points it has necessarily already been checked.
+
+This analysis defines `cfg_dominatorst` (dominator analysis) and
+`cfg_post_dominatorst` (post-dominator analysis). Run the analysis using
+`operator()(const goto_programt &)` -- it appears to be templated for other
+program types, but in truth this has to be `goto_programt &` and was only
+templated because `goto_programt` itself used to be templated.
```
Comment (tautschnig) [inline src/analyses/README.md]: No, that isn't actually the reason: the intended types can be a `goto_programt` or a `const goto_programt`. Most use cases will be perfectly happy with `const goto_programt`, except `scratch_programt`, which uses `natural_loops_templatet`, which in turn uses `cfg_dominatorst` - all on a modifiable `goto_programt`.

**Review by peterschrammel — APPROVED** [review]

File: src/analyses/README.md (hunk 1 above)
Comment (smowton) [reply src/analyses/README.md]: That would surely just cast to a const-ref?

Comment (smowton) [reply src/analyses/README.md]: Experimented with this a bit and now see why it differs -- updated the README accordingly

**Comment by smowton** [comment]
@tautschnig while investigating how `natural_loops_templatet` uses the cfg-dominators class, I got to thinking how we can fight this sort of "infectious template" when it isn't really necessary.

In the particular case where the only variation is const vs. non-const, I suggest it's almost always better to de-template the core algorithm and have it work with const references everywhere (this is both easier on the compiler and makes it clearer to the programmer that this "const-neutral" code must in fact assume everything is const), then cast the results appropriately. I've made a draft PR doing this here: https://github.com/diffblue/cbmc/compare/develop...smowton:smowton/prototype/untemplate-natural-loops

Basically the tradeoff is the core functions of `natural_loopst` being duplicated by templating (and expanded in each TU that uses them), vs. a wrapper layer, and that wrapper layer imposing an extra copy of the data structure as it has to turn `std::map<const X, std::set<const X>>` into `std::map<X, std::set<X>>` which is not guaranteed trivial.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4595fce63319ef434bb1030df74e51acde384144).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91366600

**Comment by peterschrammel** [comment]
@smowton, yes, this makes sense. Can you create a ticket for this please?

**Comment by smowton** [comment]
@peterschrammel created here https://github.com/diffblue/cbmc/pull/3432
